src/web/rustdoc.rs

```diff
@@ -286,28 +286,29 @@ pub fn rustdoc_html_server_handler(req: &mut Request) -> IronResult<Response> {
     // * If there is an exact match, but the requested crate name was corrected (dashes vs. underscores), redirect to the corrected name.
     // * If there is a semver (but not exact) match, redirect to the exact version.
     // * Otherwise, return a 404.
-    let version = if let Some(match_vers) = match_version(&mut conn, &name, url_version) {
-        match match_vers.version {
-            MatchSemver::Exact((version, _)) => {
-                // Redirect when the requested crate name isn't correct
-                if let Some(name) = match_vers.corrected_name {
-                    return redirect(&name, &version, &req_path);
-                }
+    let version = match match_version(&mut conn, &name, url_version) {
```

Comment:
To simplify this, could you use `map_err` instead? Then you wouldn't need to match on the `Err` case.
```suggestion
    let version = match match_version(&mut conn, &name, url_version).map_err(|err|  IronError::new(err, status::NotFound))?.version {
```